File: pegasus/services/optimizer-core/server/mapping.py
from __future__ import annotations

from typing import Iterable


class BidirectionalIndexMap:
    """Deterministic UUID <-> index mapping for a single solver run."""

    def __init__(self, ordered_uuids: Iterable[str]) -> None:
        self._uuid_to_index: dict[str, int] = {}
        self._index_to_uuid: list[str] = []

        for raw_uuid in ordered_uuids:
            uuid = raw_uuid.strip()
            if not uuid:
                continue
            if uuid in self._uuid_to_index:
                raise ValueError(f"duplicate uuid in mapping input: {uuid}")
            self._uuid_to_index[uuid] = len(self._index_to_uuid)
            self._index_to_uuid.append(uuid)

    @property
    def size(self) -> int:
        return len(self._index_to_uuid)

    def index_of(self, uuid: str) -> int:
        return self._uuid_to_index[uuid]

    def uuid_of(self, index: int) -> str:
        return self._index_to_uuid[index]

    def ordered(self) -> list[str]:
        return list(self._index_to_uuid)

    def as_index_map(self) -> dict[str, int]:
        return dict(self._uuid_to_index)

File: pegasus/services/optimizer-core/server-rust/src/mapping.rs
#![allow(dead_code)]

use std::collections::HashMap;

#[derive(Debug, Clone)]
pub struct BidirectionalIndexMap {
    uuid_to_index: HashMap<String, usize>,
    index_to_uuid: Vec<String>,
}

impl BidirectionalIndexMap {
    pub fn new<'a, I>(ordered_uuids: I) -> Result<Self, String>
    where
        I: IntoIterator<Item = &'a str>,
    {
        let mut uuid_to_index = HashMap::new();
        let mut index_to_uuid = Vec::new();

        for raw_uuid in ordered_uuids {
            let uuid = raw_uuid.trim();
            if uuid.is_empty() {
                continue;
            }
            if uuid_to_index.contains_key(uuid) {
                return Err(format!("duplicate uuid in mapping input: {uuid}"));
            }
            uuid_to_index.insert(uuid.to_string(), index_to_uuid.len());
            index_to_uuid.push(uuid.to_string());
        }

        Ok(Self {
            uuid_to_index,
            index_to_uuid,
        })
    }

    pub fn size(&self) -> usize {
        self.index_to_uuid.len()
    }

    pub fn index_of(&self, uuid: &str) -> Option<usize> {
        self.uuid_to_index.get(uuid).copied()
    }

    pub fn uuid_of(&self, index: usize) -> Option<&str> {
        self.index_to_uuid.get(index).map(String::as_str)
    }

    pub fn ordered(&self) -> Vec<String> {
        self.index_to_uuid.clone()
    }
}
